### PR TITLE
Add logging facade

### PR DESCRIPTION
**Pull request by sdonkov**
Adding log4j2 because it's the fastest and most advanced. Trying to remove all system outs.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,6 +33,21 @@
             <version>5.9.1</version>
             <scope>test</scope>
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-core -->
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <version>2.19.0</version>
+        </dependency>
```
Comment (dmitankin) [inline pom.xml:40]: This is the same as the one below. Self-review and all that...

File: pom.xml [hunk 2]
```diff
@@ -33,6 +33,21 @@
             <version>5.9.1</version>
             <scope>test</scope>
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-core -->
```
Comment (dmitankin) [inline pom.xml]: No need for this comment really. Same below.

File: src/main/java/com/github/sdonkov/PrefixTree.java [hunk 3]
```diff
@@ -3,8 +3,12 @@
 import java.util.ArrayList;
 import java.util.Optional;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
 public class PrefixTree {
 
+    private final static Logger logger = LogManager.getLogger(PrefixTree.class);
```
Comment (dmitankin) [inline src/main/java/com/github/sdonkov/PrefixTree.java]: `final static` fields are usually named upper-case, being constants.

File: src/main/java/com/github/sdonkov/PrefixTree.java [hunk 4]
```diff
@@ -61,21 +65,18 @@ public String toString() {
     public boolean contains(CharSequence word) {
         TraverseResult traverseResult = traverse(word);
         if (traverseResult.index == word.length() && traverseResult.node.isEndOfWord()) {
-            //todo
-            System.out.println("Tree contains this word - " + word);
+            logger.debug("Tree contains this word - " + word);
```
Comment (dmitankin) [inline src/main/java/com/github/sdonkov/PrefixTree.java]: This is really a `trace` statement (see [here](https://softwareengineering.stackexchange.com/questions/279690/why-does-the-trace-level-exist-and-when-should-i-use-it-rather-than-debug) for a discussion), but that's not a big deal. However, string concatenation is expensive and should be done only when it matters. Here, the string is always constructed while it is being used only when debug logging is turned on. Log statements must be guarded with [a check](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/src-html/org/apache/logging/log4j/Logger.html#line.2371) whether logging will be performed whenever they involve more than the invocation itself.
An alternative to the check is to use [parameterized messages](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/Logger.html#trace-java.lang.String-java.lang.Object-) (preferred) or [lambdas](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/Logger.html#trace-java.lang.String-org.apache.logging.log4j.util.Supplier...-).

**Review by dmitankin — COMMENTED** [review]
Please provide an example on how debug log statements look like now.

File: src/main/java/com/github/sdonkov/PrefixTree.java [hunk 5]
```diff
@@ -61,21 +65,24 @@ public String toString() {
     public boolean contains(CharSequence word) {
         TraverseResult traverseResult = traverse(word);
         if (traverseResult.index == word.length() && traverseResult.node.isEndOfWord()) {
-            //todo
-            System.out.println("Tree contains this word - " + word);
+            if (LOGGER.isTraceEnabled()) {
```
Comment (dmitankin) [inline src/main/java/com/github/sdonkov/PrefixTree.java]: Here and below, there's no need for the `isTraceEnabled` check now that you are no longer constructing a String instance.

`"Tree contains this word - " + word` is producing the same (or very similar) byte-code as:
```
StringBuilder sb = new StringBuilder();
sb.add("Tree contains this word - ");
sb.add(word);
sb.toString();
```
This is expensive as each `add` invocation is allocating memory (or it has been already allocated, depending on size). The problem is that this string building occurs before calling `trace`. Now, `trace` does nothing if it's not enabled, so the string building goes to waste in that case (which is the usual case).

The current variant `LOGGER.trace("Tree contains this word - {}", word);` is not allocating memory right away. It calls `trace` with a couple of parameters that are readily available - no string building is involved. Inside `trace`, the `isTraceEnabled` check is performed and if logging on this level is not enabled, no extra string is constructed (`trace` simply returns). If enough invocations of `contains` happen, the JVM compiles it to machine code. During the initial invocations, it figures out that `trace` does nothing so in the machine-code version the invocation doesn't happen at all.

If trace logging is enabled, the two are eqivalent.

File: src/test/resources/log4j2.xml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="INFO">
+    <Appenders>
+        <Console name="ConsoleAppender" target="SYSTEM_OUT">
+            <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
```
Comment (dmitankin) [inline src/test/resources/log4j2.xml]: The `%logger{36}` thing logs the fully-quallified name of the logger which is usually excessive. `%c{1}` would be a better fit for the tests.

`%logger{36}`:
```
08:56:34.965 [main] TRACE com.github.sdonkov.PrefixTree - Tree doesn't contain this word - aa
```

 `%c{1}`:
```
08:59:12.020 [main] TRACE PrefixTree - Tree doesn't contain this word - aa
```

See [here](https://logging.apache.org/log4j/2.x/manual/layouts.html#Patterns) for explanation of `%c{1}`.

**Comment by sdonkov** [comment]
Screenshot 2023-01-13 224427

**Review by dmitankin — APPROVED** [review]

**Pull request merged**